vmm/src/vm.rs

```diff
@@ -973,9 +984,41 @@ impl Vm {
             Some(arch::layout::HIGH_RAM_START),
         ) {
             Ok(entry_addr) => entry_addr,
-            Err(e) => {
-                return Err(Error::KernelLoad(e));
-            }
+            Err(e) => match e {
+                Elf(InvalidElfMagicNumber) => {
+                    // Not an ELF header - assume raw binary data / firmware
+                    let size = kernel.seek(SeekFrom::End(0)).map_err(Error::FirmwareFile)?;
+
+                    // The OVMF firmware is as big as you might expect and it's 4MiB so limit to that
+                    if size > 4 << 20 {
+                        return Err(Error::FirmwareTooLarge);
+                    }
+
+                    // Loaded at the end of the 4GiB
+                    let load_address = GuestAddress(4 << 30)
+                        .checked_sub(size)
+                        .ok_or(Error::FirmwareTooLarge)?;
+
+                    self.memory_manager
+                        .lock()
+                        .unwrap()
+                        .add_ram_region(load_address, size as usize)
+                        .map_err(Error::AllocateFirmwareMemory)?;
```

Comment:
So you're creating a RAM region at the end of the memory hole. We should be careful about that since that means the hole doesn't go as high as 4G. That means the allocator responsible for `mmio_hole_allocate()` should be aware of that.

Reply:
The MMIO device hole is only 640MiB starting at 3GiB + 256MiB. (This is identical to how TDX firmware is loaded)

Reply:
Sounds good!